CallsAndSMS/app/src/main/java/github/nisrulz/example/callsandsms/MainActivity.kt

```diff
@@ -0,0 +1,158 @@
+
+import android.Manifest
+import android.content.Intent
+import android.content.IntentFilter
+import android.content.pm.PackageManager
+import android.net.Uri
+import android.os.Bundle
+import android.text.TextUtils
+import android.widget.Toast
+import androidx.activity.result.contract.ActivityResultContracts
+import androidx.appcompat.app.AppCompatActivity
+import androidx.core.content.ContextCompat
+import github.nisrulz.example.callsandsms.databinding.ActivityMainBinding
+
+private const val TELEPHONE_PROTOCOL = "tel"
+private const val SMS_PROTOCOL = "smsto"
+
+class MainActivity : AppCompatActivity() {
+
+    companion object {
+        private val REQUIRED_PERMISSIONS = arrayOf(
+            Manifest.permission.CALL_PHONE,
+            Manifest.permission.SEND_SMS,
+            Manifest.permission.RECEIVE_SMS,
+            Manifest.permission.READ_PHONE_STATE
+        )
+    }
+
+    private lateinit var phoneCallStateReceiver: PhoneCallStateReceiver
+    private lateinit var smsReceiver: SMSReceiver
+
+    private lateinit var binding: ActivityMainBinding
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityMainBinding.inflate(layoutInflater)
+        binding.apply {
+            setContentView(root)
+            checkForPermissions()
+            setOnClickListeners()
+        }
+    }
+
+    override fun onResume() {
+        super.onResume()
+        val phoneStateFilter = IntentFilter("android.intent.action.PHONE_STATE")
+        val smsReceiverFilter = IntentFilter("android.provider.Telephony.SMS_RECEIVED")
+        phoneCallStateReceiver = PhoneCallStateReceiver()
+        smsReceiver = SMSReceiver()
+        registerReceiver(phoneCallStateReceiver, phoneStateFilter)
+        registerReceiver(smsReceiver, smsReceiverFilter)
+    }
+
+    override fun onPause() {
+        super.onPause()
+        unregisterReceiver(phoneCallStateReceiver)
+        unregisterReceiver(smsReceiver)
+    }
+
+    private fun setOnClickListeners() {
+        binding.apply {
+            btnDial.setOnClickListener {
+                val phoneNo = etPhoneNo.text.toString()
+                if (!TextUtils.isEmpty(phoneNo)) {
+                    val dialUri = createUriForProtocol(TELEPHONE_PROTOCOL, phoneNo)
+                    startActivity(Intent(Intent.ACTION_DIAL, Uri.parse(dialUri)))
+                } else {
+                    showToast("Enter a phone number")
+                }
+            }
+            btnCall.setOnClickListener {
+                val phoneNo = etPhoneNo.text.toString()
+                if (!TextUtils.isEmpty(phoneNo)) {
+                    val dialUri = createUriForProtocol(TELEPHONE_PROTOCOL, phoneNo)
+                    // Requires Permission to be declared in manifest
+                    // <uses-permission android:name="android.permission.CALL_PHONE"/>
+                    // Then check and request for permission during runtime
```

Comment:
I tend to also use `@RequirePermission` in code to  have IDE also help.

https://developer.android.com/reference/androidx/annotation/RequiresPermission

Reply:
Added